jsettlers.logic/src/jsettlers/ai/highlevel/AiStatistics.java

```diff
@@ -0,0 +1,612 @@
+		Integer playerIdInteger = new Integer(playerId);
+		if (numberOfNotOccupiedTowers.get(playerIdInteger) == null) {
+			return 0;
+		}
+		return numberOfNotOccupiedTowers.get(playerIdInteger);
+	}
+
+	public boolean blocksWorkingAreaOfOtherBuilding(ShortPoint2D point) {
+		return pointIsBlocked((point.x), (short) (point.y - 12))
+				|| pointIsBlocked((short) (point.x - 5), (short) (point.y - 12))
+				|| pointIsBlocked((short) (point.x - 10), (short) (point.y - 12))
+				|| pointIsBlocked((point.x), (short) (point.y - 6))
+				|| pointIsBlocked((short) (point.x - 5), (short) (point.y - 6))
+				|| pointIsBlocked((short) (point.x - 10), (short) (point.y - 6));
+	}
+
+	private boolean pointIsBlocked(short x, short y) {
+		IBuilding building = objectsGrid.getBuildingOn(x, y);
+		if (building != null && (building.getBuildingType() == LUMBERJACK || building.getBuildingType() == FARM
+				|| building.getBuildingType() == WINEGROWER)) {
+			return true;
+		}
+		return false;
+	}
+
+	public boolean southIsFreeForPlayer(ShortPoint2D point, byte playerId) {
+		return pointIsFreeForPlayer(point.x, (short) (point.y + 12), playerId)
+				&& pointIsFreeForPlayer((short) (point.x + 5), (short) (point.y + 12), playerId)
+				&& pointIsFreeForPlayer((short) (point.x + 10), (short) (point.y + 12), playerId)
+				&& pointIsFreeForPlayer(point.x, (short) (point.y + 6), playerId)
+				&& pointIsFreeForPlayer((short) (point.x + 5), (short) (point.y + 6), playerId)
+				&& pointIsFreeForPlayer((short) (point.x + 10), (short) (point.y + 6), playerId);
+	}
+
+	private boolean pointIsFreeForPlayer(short x, short y, byte playerId) {
+		return partitionsGrid.getPlayerIdAt(x, y) == playerId && !objectsGrid.isBuildingAreaAt(x, y) && !flagsGrid.isProtected(x, y)
+				&& landscapeGrid.areAllNeighborsOf(x, y, 0, 2, ELandscapeType.GRASS, ELandscapeType.EARTH);
+	}
+
+	public Movable getNearestSwordsmanOf(ShortPoint2D targetPosition, byte playerId) {
+		List<ShortPoint2D> soldierPositions = getMovablePositionsByTypeForPlayer(SWORDSMAN_L3, playerId);
+		if (soldierPositions.size() == 0) {
+			soldierPositions = getMovablePositionsByTypeForPlayer(SWORDSMAN_L2, playerId);
+		}
+		if (soldierPositions.size() == 0) {
+			soldierPositions = getMovablePositionsByTypeForPlayer(SWORDSMAN_L1, playerId);
+		}
+		if (soldierPositions.size() == 0) {
+			return null;
+		}
+
+		ShortPoint2D nearestSoldierPosition = detectNearestPointFromList(targetPosition, soldierPositions);
+		return movableGrid.getMovableAt(nearestSoldierPosition.x, nearestSoldierPosition.y);
+	}
+
+	public ShortPoint2D detectNearestPointFromList(ShortPoint2D referencePoint, List<ShortPoint2D> points) {
+		ShortPoint2D nearestPoint = null;
+		double nearestPointDistance = Double.MAX_VALUE;
+		for (ShortPoint2D point : points) {
+			double currentPointDistance = referencePoint.calculateDistanceTo(point);
+			if (nearestPoint == null || currentPointDistance < nearestPointDistance) {
```

Comment:
```currentPointDistance < nearestPointDistance```is only true for the first iteration. 2 lines below nearestPointDistance is modified.

```
int nearestPointDistance = Integer.MAX_VALUE;
		for (ShortPoint2D point : points) {
			int currentPointDistance = referencePoint.getOnGridDistTo(point);
			if (nearestPoint == null || currentPointDistance < nearestPointDistance) {
				nearestPoint = point;
				nearestPointDistance = currentPointDistance;
			}
		}
```